src/Initializer/ParameterDB.cpp

```diff
@@ -178,119 +217,220 @@ easi::Query seissol::initializers::FaultGPGenerator::generate() const {
+      adapter.addBindingPoint("c13", &AnisotropicMaterial::c13);
+      adapter.addBindingPoint("c14", &AnisotropicMaterial::c14);
+      adapter.addBindingPoint("c15", &AnisotropicMaterial::c15);
+      adapter.addBindingPoint("c16", &AnisotropicMaterial::c16);
+      adapter.addBindingPoint("c22", &AnisotropicMaterial::c22);
+      adapter.addBindingPoint("c23", &AnisotropicMaterial::c23);
+      adapter.addBindingPoint("c24", &AnisotropicMaterial::c24);
+      adapter.addBindingPoint("c25", &AnisotropicMaterial::c25);
+      adapter.addBindingPoint("c26", &AnisotropicMaterial::c26);
+      adapter.addBindingPoint("c33", &AnisotropicMaterial::c33);
+      adapter.addBindingPoint("c34", &AnisotropicMaterial::c34);
+      adapter.addBindingPoint("c35", &AnisotropicMaterial::c35);
+      adapter.addBindingPoint("c36", &AnisotropicMaterial::c36);
+      adapter.addBindingPoint("c44", &AnisotropicMaterial::c44);
+      adapter.addBindingPoint("c45", &AnisotropicMaterial::c45);
+      adapter.addBindingPoint("c46", &AnisotropicMaterial::c46);
+      adapter.addBindingPoint("c55", &AnisotropicMaterial::c55);
+      adapter.addBindingPoint("c56", &AnisotropicMaterial::c56);
+      adapter.addBindingPoint("c66", &AnisotropicMaterial::c66);
     }                                                               
     
     template<class T>
-    void MaterialParameterDB<T>::evaluateModel(std::string const& fileName, QueryGenerator const& queryGen) {
+    void MaterialParameterDB<T>::evaluateModel(std::string const& fileName, QueryGenerator const * const queryGen) {
       easi::Component* model = loadEasiModel(fileName);
-      easi::Query query = queryGen.generate();
+      easi::Query query = queryGen->generate();
+      const unsigned numPoints = query.numPoints();
+
+      std::vector<T> materialsFromQuery(numPoints);
+      easi::ArrayOfStructsAdapter<T> adapter(materialsFromQuery.data());
+      MaterialParameterDB<T>().addBindingPoints(adapter);
+      model->evaluate(query, adapter);
+
+      // Only use homogenization when ElementAverageGenerator has been supplied
+      if (const ElementAverageGenerator* gen = dynamic_cast<const ElementAverageGenerator*>(queryGen)) {
+        const unsigned numElems = numPoints / NUM_QUADPOINTS;
+        std::array<double, NUM_QUADPOINTS> quadratureWeights{ gen->getQuadratureWeights() };
+
+        // Compute homogenized material parameters for every element in a specialization for the particular material
+        #pragma omp parallel for
+        for (unsigned elementIdx = 0; elementIdx < numElems; ++elementIdx) {
+          m_materials->at(elementIdx) = this->computeAveragedMaterial(elementIdx, quadratureWeights, materialsFromQuery); 
+        }
+      } else {
+      // Usual behavior without homogenization
+        for (unsigned i = 0; i < numPoints; ++i) {
+          m_materials->at(i) = T(materialsFromQuery[i]);
+        }
+      }
+      delete model;
```

Comment:
delete query?

Reply:
The query is not deleted in all of the existing evaluateModel functions as well. To maintain consistency, it would need to be added everywhere. This affects parts of the file I did not code so I won't change it.

Reply:
I see.